### PR TITLE
Data flow: Introduce `ApproxContent` in a new pruning stage between stages 2 and 3

### PR DESCRIPTION
**Pull request by hvitved**
When adding [flow through constructors for Ruby](https://github.com/github/codeql/pull/10714), we saw a large increase in tuple counts:

<details><summary>Tuple counts on `instructure/canvas-lms` before this PR:</summary>

\# | n | stage | nodes | fields | conscand | states | tuples | config
-- | -- | -- | -- | -- | -- | -- | -- | --
1 | 10 | 1 Fwd | 1383610 | 9298 | -1 | 1 | 1434561 | RegExpConfiguration
2 | 15 | 1 Rev | 484212 | 7631 | -1 | 1 | 504458 | RegExpConfiguration
3 | 20 | 2 Fwd | 83935 | 1951 | 2709 | 1 | 175206 | RegExpConfiguration
4 | 25 | 2 Rev | 64808 | 1639 | 2278 | 1 | 82827 | RegExpConfiguration
5 | 30 | 3 Fwd | 32603 | 999 | 106977 | 1 | 38562854 | RegExpConfiguration
6 | 35 | 3 Rev | 21479 | 875 | 76399 | 1 | 6520738 | RegExpConfiguration
7 | 40 | 4 Fwd | 17307 | 586 | 5894 | 1 | 488804 | RegExpConfiguration
8 | 45 | 4 Rev | 9077 | 313 | 3626 | 1 | 117635 | RegExpConfiguration
9 | 50 | 5 Fwd | 8261 | 299 | 1297 | 1 | 54664 | RegExpConfiguration
10 | 55 | 5 Rev | 203 | 12 | 14 | 1 | 211 | RegExpConfiguration

</details>

The explosion in tuple counts in pruning stage 3 happens because we go from using a Boolean to represent access paths (`true` <=> non-empty), to representing access paths by their head, which means that once we see a read of content `c`, we will often end up with a large number of resulting access paths (any head which is seen prior to a store into content `c`).

This PR attempts to reduce the tuple explosion, by adding a new pruning stage in between stages 2 and 3, where we represent access paths not by the head of the access path, but by an _approximated_ head. An approximated head is defined in terms of a new language-specific type:

```ql
/** An approximated `Content`. */
class ContentApprox;

/** Gets an approximated value for content `c`. */
ContentApprox getContentApprox(Content c);
```

That is, it is up to each language to decide exactly how to approximate `Content`s (and therefore how access path heads are approximated in the new pruning stage).

For Ruby, this means a significant reduction in tuple counts (at the cost of an extra pruning stage):

<details><summary>Tuple counts on `instructure/canvas-lms` with this PR:</summary>

\# | n | stage | nodes | fields | conscand | states | tuples | config
-- | -- | -- | -- | -- | -- | -- | -- | --
1 | 10 | 1 Fwd | 1383610 | 9298 | -1 | 1 | 1434561 | RegExpConfiguration
2 | 15 | 1 Rev | 484212 | 7631 | -1 | 1 | 504458 | RegExpConfiguration
3 | 20 | 2 Fwd | 83925 | 1950 | 2708 | 1 | 175124 | RegExpConfiguration
4 | 25 | 2 Rev | 64798 | 1638 | 2277 | 1 | 82984 | RegExpConfiguration
5 | 30 | 3 Fwd | 33673 | 1157 | 33635 | 1 | 2396032 | RegExpConfiguration
6 | 35 | 3 Rev | 23052 | 985 | 19149 | 1 | 536699 | RegExpConfiguration
7 | 40 | 4 Fwd | 21063 | 865 | 57186 | 1 | 13210734 | RegExpConfiguration
8 | 45 | 4 Rev | 20346 | 839 | 54535 | 1 | 5501937 | RegExpConfiguration
9 | 50 | 5 Fwd | 17161 | 583 | 5884 | 1 | 488752 | RegExpConfiguration
10 | 55 | 5 Rev | 9077 | 313 | 3626 | 1 | 117635 | RegExpConfiguration
11 | 60 | 6 Fwd | 8261 | 299 | 1297 | 1 | 54524 | RegExpConfiguration
12 | 65 | 6 Rev | 203 | 12 | 14 | 1 | 211 | RegExpConfiguration

</details>

For all languages except C#, Java, and Ruby, `ContentApprox` is for now simply an alias for `Unit`, which means that the new pruning stage will be similar to stage 2, except for languages that do type pruning, since types are also tracked in this new stage.

### Future work

There are different directions for future work:
- Language maintainers implement their own `ContentApprox`.
- We no longer make `ContentApprox` language-specific, and instead attempt to derive suitable approximations inside the data flow library.

Even if we eventually decide to pursue the second approach, this PR still makes sense, as it unblocks Ruby here and now.

### CODE REVIEW

**Comment by MathiasVP** [comment]
Uh, this looks very relevant for the stuff we're doing in https://github.com/github/codeql/pull/10817 as well! Do let me know once this PR is ready for testing, and I'd love to try it out.

**Comment by hvitved** [comment]
> Uh, this looks very relevant for the stuff we're doing in #10817 as well! Do let me know once this PR is ready for testing, and I'd love to try it out.

I have updated the PR with a description now. Let me know if it helps with your work as well.

File: ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -1364,3 +1371,58 @@ predicate additionalLambdaFlowStep(Node nodeFrom, Node nodeTo, boolean preserves
...
+    or
+    exists(Content c |
+      this = TNonElementContentApprox(c) and
+      result = c.toString()
+    )
+  }
+}
+
+/**
+ * Gets a string for approximating know element indices.
```
Comment (MathiasVP) [inline ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowPrivate.qll:1397]:
```suggestion
 * Gets a string for approximating known element indices.
```

**Review by aschackmull — APPROVED** [review]
Looks reasonable. Follow-up tweaks discussed offline.

**Pull request merged**